File: Metodos de aproximacion iterativos/Metodo de la Falsa Posicion/Falsa_Posicion.m
function Falsa_Posicion
  pkg load symbolic
  a=0;
  b=pi;
  tol=1e-10;
  iterMax=2500;
  f="cos(2*x)^2-x^2";
  [xk,error,k]=falsa(f,a,b,tol,iterMax);
  if(xk!=inf)
    display(["El valor aproximado por el metodo es de ",num2str(xk)," con un error de ",num2str(error)," en ",num2str(k)," iteraciones"])
  endif
end

function[xk,error,k]=falsa(f,a,b,tol,iterMax)
  f=sym(f);
  f=matlabFunction(f);
  xk=inf;
  error=inf;
  k=0;
  if(f(a)*f(b)<0)
    for i=0:iterMax
      if(f(b)-f(a)==0)
        disp("Indefinicion en alguna de las iteraciones, se retornan los valores hasta dicha iteracion");
        return
      endif
      xk=b-((b-a)/(f(b)-f(a)))*f(b);
      error=f(xk);
      
      if(abs(error)<=tol)
        return
      endif
      
      if(error*f(a)<0)
        b=xk;
      elseif(error*f(b)<0)
        a=xk;
      else
        disp("Ninguno de los nuevos intervalos cumple con el teorema de Bolzano");
        xk=inf;
        return
      endif
      k=k+1;
      
    endfor
    return
  endif
  
  disp("La funcion provista no cumple el teorema de Bolzano en el intervalo especificado");
  return
end

File: Metodos de aproximacion iterativos/Metodo de la Falsa Posicion/Falsa Posicion.py
import numpy as np
from sympy import *
from math import *

from sympy.calculus.util import continuous_domain

#Metodo para determinar si una funcion es contiua en el intervalo de [a,b]
'''Entradas: f=Funcion a evaluar
             a=Limite inferior del intervalo en el que se debe evaluar la continuidad de la funcion
             b=limite superior del intervalo en el que se debe evaluar la continuidad de la funcion
   Salidas: Booleano que inidica si la funcion es continua sobre el intervalo [a,b]'''

def isContinuous(f,a,b):
    f=sympify(f)
    x=symbols('x')
    var=continuous_domain(f,x,Interval(a,b))
    return Interval(a,b)==var

def falsa_posicion(f,a,b,tol,iterMax):
    if(isContinuous(f,a,b)):
        x=symbols('x')
        f=lambdify(x,f)
        if(f(a)*f(b)<0):
            xk,error,k=0,0,0
            while(k<iterMax):
                if(f(b)-f(a)==0):
                    print("Indefinicion en alguna de las iteraciones, se retornan los valores hasta dicha iteracion")
                    break
                xk = b-((b - a)/(f(b) - f(a)))* f(b)
                error=f(xk)
                if(abs(error)<=tol):
                    break
                if(error*f(a)<0):
                    b=xk
                elif(error*f(b)<0):
                    a=xk
                else:
                    print("Ninguno de los nuevos intervalos cumple con el teorema de Bolzano")
                    return (None,None,None)
                k+=1
            return(xk,error,k)
        else:
            print("La funcion no cumple con el teorema de Bolzano")
            return (None,None,None)
    else:
        print("La funcion provista no es continua en el intervalo espcificado")
        return (None,None,None)
f="sin(x)**2+x**2-1"
valores=falsa_posicion(f,0,2,10**-10,2500)
if(valores[0]!=None):
    print("La aproximacion encontrada es x= "+str(valores[0])+" con un error de "+str(valores[1])+" en "+str(valores[2])+" iteraciones")
